Library/Homebrew/cask/cask.rb

```diff
@@ -151,18 +157,38 @@ def caskfile_only?
 
     sig { returns(T.nilable(Time)) }
     def install_time
-      _, time = timestamped_versions.last
-      return unless time
-
-      Time.strptime(time, Metadata::TIMESTAMP_FORMAT)
+      time = installed_caskfile&.dirname&.dirname&.basename&.to_s
+      Time.strptime(time, Metadata::TIMESTAMP_FORMAT) if time
     end
 
+    sig { returns(T.nilable(Pathname)) }
     def installed_caskfile
-      installed_version = timestamped_versions.last
-      caskfile_dir = metadata_main_container_path.join(*installed_version, "Casks")
-      return caskfile_dir.join("#{token}.json") if caskfile_dir.join("#{token}.json").exist?
+      installed_caskroom_path = caskroom_path
+      installed_token = token
+
+      # Check if the cask is installed with an old name.
+      old_tokens&.each do |old_token|
+        if (old_caskroom_path = Caskroom.path.join(old_token)).directory?
+          installed_caskroom_path = old_caskroom_path
+          installed_token = old_token
+          break
+        end
+      end
+
+      installed_version = timestamped_versions(caskroom_path: installed_caskroom_path).last
+      return unless installed_version
+
+      caskfile_dir = metadata_main_container_path(caskroom_path: installed_caskroom_path)
+                     .join(*installed_version, "Casks")
 
-      caskfile_dir.join("#{token}.rb")
+      ["json", "rb"]
+        .map { |ext| caskfile_dir.join("#{installed_token}.#{ext}") }
+        .find(&:exist?)
+    end
+
+    sig { returns(T.nilable(String)) }
+    def installed_version
+      installed_caskfile&.dirname&.dirname&.dirname&.basename&.to_s
```

Comment:
Should `#installed_caskfile` and `#installed_version` get memoized here?

Reply:
No. There is actually a test which was silently broken because the installed version was cached.

Reply:
Not sure I understand. The version wasn't cached before, right? Which test was silently broken?

Reply:
I was thinking of a different issue. The issue was that previously `installed?` counted timestamped directories which don't contain a `.rb` or `.json`. That was a issue caused somewhere in `Installer`.

I still don't think it's a good idea to cache installed versions, since it'll probably break “install” and “uninstall” tests.